### PR TITLE
media api initial implementation

### PR DESCRIPTION
**Pull request by Ansu-s**
Added GET,POST,PUT,DELETE methods for Media-API

### CODE REVIEW

File: src/main/java/com/example/bookstoreapp/services/impl/MediaServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Service
+public class MediaServiceImpl implements MediaService {
+
+    @Autowired
+    MediaRepository mediaRepository;
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/services/impl/MediaServiceImpl.java:19]: Access modifier required

File: src/main/java/com/example/bookstoreapp/services/impl/MediaServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                return new Media().fromEntity(savedMedia.get());
+            } else throw new AppRuntimeException("No item found with the id " + id);
+        } catch (Exception e) {
+            throw new AppRuntimeException("No such item is present with this id " + id);
+        }
+    }
+
+    @Override
+    public Media updateMedia(Long id, Media media) {
+        Media sevedMedia = findById(id);
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/services/impl/MediaServiceImpl.java:42]: Spell error

File: src/main/java/com/example/bookstoreapp/services/impl/MediaServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            }
+        } catch (Exception e) {
+            throw new AppRuntimeException("Error while deleting media with id " + id);
+        }
+    }
+
+    @Override
+    public List<Media> allMedia() {
+        Iterable<MediaEntity> entities = mediaRepository.findAll();
+        List<Media> mediaList = new ArrayList<>();
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/services/impl/MediaServiceImpl.java:67]: Do we have any other approach already implemented in our codebase?

**Review by rkpatra201 — COMMENTED** [review]
Reviewed

**Comment by Ansu-s** [comment]
I have created a endpoint which will return all the media details for particular Item and all the media elements are sorted according to their sequenceId

  
image

**Comment by Ansu-s** [comment]
Removed the ItemMedia API. Please review the final version.

**Comment by Ansu-s** [comment]
Added test cases for controller and services with 100% coverage  
image

File: src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java [hunk 4]
```diff
@@ -12,159 +12,164 @@
...
 import com.example.bookstoreapp.utils.IdGenerator;
 import jakarta.persistence.EntityManager;
-import jakarta.persistence.EntityTransaction;
-import jakarta.persistence.criteria.CriteriaBuilder;
-import jakarta.persistence.criteria.CriteriaDelete;
-import jakarta.persistence.criteria.CriteriaQuery;
-import jakarta.persistence.criteria.CriteriaUpdate;
-import jakarta.persistence.criteria.Predicate;
-import jakarta.persistence.criteria.Root;
+import jakarta.persistence.criteria.*;
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java:15]: we dont need wildcard import statement

File: src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java [hunk 5]
```diff
@@ -12,159 +12,164 @@
...
+            Predicate statePredicate = criteriaBuilder.equal(root.get("state"), cartItemState.getWeight());
+            predicateList.add(statePredicate);
+        }
+        criteriaQuery.where(predicateList.toArray(new Predicate[0]));
+        // select * from cart_item where cart_id=?
+        // select * from cart_item where cart_id=? and state=?
+
+        return entityManager.createQuery(criteriaQuery).getResultList();
     }
-    // increase the quantity
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java:65]: why does it looks like below lines cleaned up?

File: src/main/java/com/example/bookstoreapp/models/CartItem.java [hunk 6]
```diff
@@ -7,26 +7,27 @@
 @Data
 public class CartItem extends CatalogItem {
 
-  private Long cartId;
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/models/CartItem.java:10]: This class re-written which is not required I think,

File: src/main/java/com/example/bookstoreapp/models/CatalogItem.java [hunk 7]
```diff
@@ -10,20 +10,23 @@
...
-    this.setPrice(entity.getPrice());
-    this.setStockQuantity(entity.getStockQuantity());
-    return this;
-  }
+    public CatalogItem fromEntity(CatalogItemEntity entity) {
+        this.setId(entity.getId());
+        this.setName(entity.getName());
+        this.setPrice(entity.getPrice());
+        this.setStockQuantity(entity.getStockQuantity());
+        this.setItemLimit(entity.getItemLimit());
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/models/CatalogItem.java:29]: item limit changes not required in this pull request, you can stash the commit to separate branch and create another pull requests for item limit.

File: src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java [hunk 8]
```diff
@@ -12,159 +12,164 @@
...
+
+            return new CartItem().fromEntity(newCartItemEntity, catalogItem);
+        }
+
+        /* An item can have two states : One in which it has a limit & the other in which it has no limit
+        if item doesn't have any limit = In database (item_limit = 0)
+        if item has limit = In database item_limit is set to the limit count
+         */
+        // When there is no item limit
+        if (catalogItem.getItemLimit() == 0) {
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java:64]: These if statements looks like duplicate to me. 

```java
if (catalogItem.getItemLimit() != 0) {
            if (savedCartItemEntity.getQuantity() >= catalogItem.getItemLimit()){
                throw new AppRuntimeException("You have reached the maximum limit for this item");
     }
}

savedCartItemEntity.setQuantity(savedCartItemEntity.getQuantity() + 1);
            savedCartItemEntity = cartItemEntityRepository.save(savedCartItemEntity);

```

File: src/main/java/com/example/bookstoreapp/models/CatalogItem.java [hunk 9]
```diff
@@ -10,20 +10,23 @@
 @NoArgsConstructor
 public class CatalogItem {
 
-  private Long id;
```
Comment (rkpatra201) [inline src/main/java/com/example/bookstoreapp/models/CatalogItem.java:13]: This class also re-written which is not required, just for itemLimit field.

This happens when you select all and do formatting. You should format only the code which is newly added.

**Comment by rkpatra201** [comment]
Good work. @Ansu-s . Just separate item limit functionality using stash or cherry-pick from media PR.

**Review by rkpatra201 — COMMENTED** [review]
Reviewed

File: src/main/java/com/example/bookstoreapp/models/CatalogItem.java (hunk 9 above)
Comment (Ansu-s) [reply src/main/java/com/example/bookstoreapp/models/CatalogItem.java:13]: ok I understod what is causing these seem as newly added code ,I will keep this in consideration to avoid post this.

File: src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java (hunk 4 above)
Comment (Ansu-s) [reply src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java:15]: done

File: src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java (hunk 5 above)
Comment (Ansu-s) [reply src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java:65]: yes I have added code clean up

File: src/main/java/com/example/bookstoreapp/models/CatalogItem.java (hunk 7 above)
Comment (Ansu-s) [reply src/main/java/com/example/bookstoreapp/models/CatalogItem.java:29]: crated a separated pull request for the issue

File: src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java (hunk 8 above)
Comment (Ansu-s) [reply src/main/java/com/example/bookstoreapp/services/impl/CartItemServiceImpl.java:64]: Applied this into the the code

**Review by rkpatra201 — APPROVED** [review]

**Comment by rkpatra201** [comment]
As this pull request is first submitted so I merged it. cc @Ansu-s / @Subhransupanda2000
